File: sdk-python/tests/iwfcompat/set_attributes_flow.py
# Portions of this file are derived from indeedeng/iwf-java-sdk.
# Those portions are licensed under the Apache License, Version 2.0.
# See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
#
# Modifications Copyright (c) 2026 Super Durable, Inc.
#
# Modifications are licensed under the Super Durable Source License 1.0.
# Third-Party Materials remain under the Apache License, Version 2.0.
# See LICENSE and LEGACY_NOTICES.md.

from datetime import datetime

from dex import (
    Attribute,
    AttributeIndex,
    AttributeMap,
    Flow,
    IndexType,
    PersistenceSchema,
    StepDef,
)

from .shared import CompleteStringStep


class SetAttributesFlow(Flow[str]):
    def __init__(self) -> None:
        self.data = Attribute("data", str)
        self.data_map = AttributeMap("data-map", str)
        self.keyword = Attribute(
            "keyword",
            str,
            AttributeIndex(IndexType.KEYWORD),
        )
        self.text = Attribute(
            "text",
            str,
            AttributeIndex(IndexType.FULL_TEXT),
        )
        self.decimal = Attribute(
            "double",
            float,
            AttributeIndex(IndexType.DOUBLE),
        )
        self.integer = Attribute(
            "int",
            int,
            AttributeIndex(IndexType.INT),
        )
        self.bool = Attribute(
            "bool",
            bool,
            AttributeIndex(IndexType.BOOL),
        )
        self.keywords = Attribute[tuple[str, ...]](
            "keywords",
            tuple,
            AttributeIndex(IndexType.KEYWORD_ARRAY),
        )
        self.datetime = Attribute(
            "datetime",
            datetime,
            AttributeIndex(IndexType.DATETIME),
        )
        self.start = CompleteStringStep()

    def get_steps(self) -> tuple[StepDef, ...]:
        return (StepDef.start_step(self.start),)

    def get_persistence_schema(self) -> PersistenceSchema:
        return PersistenceSchema(
            attributes=(
                self.data,
                self.data_map,
                self.keyword,
                self.text,
                self.decimal,
                self.integer,
                self.bool,
                self.keywords,
                self.datetime,
            )
        )

File: sdk-typescript/test/iwfcompat/set_attributes_flow.ts
// Portions of this file are derived from indeedeng/iwf-java-sdk.
// Those portions are licensed under the Apache License, Version 2.0.
// See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
//
// Modifications Copyright (c) 2026 Super Durable, Inc.
//
// Modifications are licensed under the Super Durable Source License 1.0.
// Third-Party Materials remain under the Apache License, Version 2.0.
// See LICENSE and LEGACY_NOTICES.md.

import {
  Attribute,
  AttributeMap,
  IndexType,
  StepList,
  booleanCodec,
  doubleCodec,
  stringCodec,
  type Flow,
  type PersistenceSchema,
} from "../../src/index.js";

import { CompleteStringStep, dateCodec, stringArrayCodec } from "./shared.js";

export class SetAttributesFlow implements Flow<string> {
  public readonly data = new Attribute("data", stringCodec);
  public readonly dataMap = new AttributeMap("data-map", stringCodec);
  public readonly keyword = new Attribute("keyword", stringCodec, {
    type: IndexType.KEYWORD,
  });
  public readonly text = new Attribute("text", stringCodec, {
    type: IndexType.FULL_TEXT,
  });
  public readonly decimal = new Attribute("double", doubleCodec, {
    type: IndexType.DOUBLE,
  });
  public readonly integer = new Attribute("int", doubleCodec, {
    type: IndexType.INT,
  });
  public readonly bool = new Attribute("bool", booleanCodec, {
    type: IndexType.BOOL,
  });
  public readonly keywords = new Attribute("keywords", stringArrayCodec, {
    type: IndexType.KEYWORD_ARRAY,
  });
  public readonly datetime = new Attribute("datetime", dateCodec, {
    type: IndexType.DATETIME,
  });
  private readonly start = new CompleteStringStep();

  public getFlowType(): string {
    return "SetAttributesFlow";
  }

  public getSteps() {
    return StepList.startStep(this.start);
  }

  public getPersistenceSchema(): PersistenceSchema {
    return {
      attributes: [
        this.data,
        this.dataMap,
        this.keyword,
        this.text,
        this.decimal,
        this.integer,
        this.bool,
        this.keywords,
        this.datetime,
      ],
    };
  }
}
